pkg/reconciler/apibinding/name_conflict_checker.go

```diff
@@ -0,0 +1,131 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apibinding
+
+import (
+	"fmt"
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+	apislisters "github.com/kcp-dev/kcp/pkg/client/listers/apis/v1alpha1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	crdlister "k8s.io/apiextensions-apiserver/pkg/client/listers/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/clusters"
+)
+
+type nameConflictChecker struct {
+	apiBindingLister        apislisters.APIBindingLister
+	apiExportLister         apislisters.APIExportLister
+	apiResourceSchemaLister apislisters.APIResourceSchemaLister
+	crdLister               crdlister.CustomResourceDefinitionLister
+}
+
+func (ncc *nameConflictChecker) getBoundCRDs(clusterName logicalcluster.LogicalCluster) ([]*apiextensionsv1.CustomResourceDefinition, error) {
+	var ret []*apiextensionsv1.CustomResourceDefinition
+
+	apiBindings, err := ncc.apiBindingLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, apiBinding := range apiBindings {
+		if apiBinding.ClusterName != clusterName.String() {
+			continue
+		}
+
+		apiExportClusterName, err := getAPIExportClusterName(apiBinding)
+		if err != nil {
+			return nil, err
+		}
+
+		apiExportKey := clusters.ToClusterAwareKey(apiExportClusterName, apiBinding.Spec.Reference.Workspace.ExportName)
+		apiExport, err := ncc.apiExportLister.Get(apiExportKey)
+		if err != nil {
+			return nil, err
+		}
+
+		boundSchemaUIDs := sets.NewString()
+		for _, boundResource := range apiBinding.Status.BoundResources {
+			boundSchemaUIDs.Insert(boundResource.Schema.UID)
+		}
+
+		for _, schemaName := range apiExport.Spec.LatestResourceSchemas {
+			schemaKey := clusters.ToClusterAwareKey(apiExportClusterName, schemaName)
+			schema, err := ncc.apiResourceSchemaLister.Get(schemaKey)
+			if err != nil {
+				return nil, err
+			}
+
+			if !boundSchemaUIDs.Has(string(schema.UID)) {
+				continue
+			}
+
+			crdKey := clusters.ToClusterAwareKey(ShadowWorkspaceName, string(schema.UID))
+			crd, err := ncc.crdLister.Get(crdKey)
+			if err != nil {
+				return nil, err
+			}
+
+			ret = append(ret, crd)
+		}
+	}
+
+	return ret, nil
+}
+
+func (ncc *nameConflictChecker) checkForConflicts(crd *apiextensionsv1.CustomResourceDefinition, clusterName logicalcluster.LogicalCluster) error {
+	existingNames := sets.NewString()
+	existingKinds := sets.NewString()
+
+	crds, err := ncc.getBoundCRDs(clusterName)
+	if err != nil {
+		return fmt.Errorf("error checking for naming conflicts for cluster %s: error getting CRDs: %w", clusterName, err)
+	}
+
+	for _, crd := range crds {
+		existingNames.Insert(crd.Status.AcceptedNames.Plural)
```

Comment:
(No action required) Hmm, so each of these 3 field values not only has to be unique across the same field in different crds but also across the other 2 fields in different crds?

Reply:
Yes - this is a reimplementation of the logic in https://github.com/kubernetes/kubernetes/blob/00bd91e558612de9b89b4c76c7a42121ac2835c6/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/status/naming_controller.go#L123.